### ISSUE TITLE
Can't seem to get Revise to work in vscode

### ISSUE THREAD

**Issue by clintonTE** [timholy/Revise.jl#325]
I'm probably doing something wrong here, but I'm not sure what it is. Hard to do a traditional MWE in this case, but I'll write down the simplest procedure I can think of. 

First do the following in some random directory:

```Julia
] generate mwe
] activate .
``` 

Edit mwe.jl in the src directory to read:

```Julia
module mwe

function func()
  println("hello")
end

end # module
```

At this point, open vscode if it is not already open. Then in the root directory (using vscode), create a file called mwescripts.jl with the following:

```Julia
using Revise, Pkg

if !isdefined(Main, :FIRST_RUN)
    Pkg.develop(PackageSpec(path=pwd()))
    FIRST_RUN = false
end

using mwe

mwe.func()
```

Running mwescripts.jl in a new repl gives:

```
hello
```

So far, so good. Then make the following change to the mwe.jl file

```Julia
module mwe

function func()
  println("hello again")
end

end # module
```

running mwescripts.jl again gives:
```
hello
```

Of course, re-starting the Repl gives the correct answer.

Running Windows 10, Julia 1.1.1. Revise passed tests.

I suspect this is a user error rather than a bug, but any help would be appreciated.

**Comment by clintonTE** [comment]
Also for good measure, here are the results after following the debugging instructions:

```Julia
Revise.LogRecord(Debug, DeleteMethod, Action, Revise_4ac0f476, "C:\Users\Clinton\.julia\packages\Revise\agmgx\src\Revise.jl", 226, (time=1.564013467903e9, deltainfo=(Tuple{typeof(mwe.func)}, MethodSummary(:func, :mwe, Symbol("C:\\Users\\Clinton\\Dropbox\\Projects\\mwe\\src\\mwe.jl"), 4, Tuple{typeof(mwe.func)}))))
Revise.LogRecord(Debug, Eval, Action, Revise_9147188b, "C:\Users\Clinton\.julia\packages\Revise\agmgx\src\Revise.jl", 276, (time=1.56401346808e9, deltainfo=(mwe, :(function func()
      #= C:\Users\Clinton\Dropbox\Projects\mwe\src\mwe.jl:4 =#
      println("hello again")
  end))))
```

**Comment by timholy** [comment]
I love VSCode but haven't used it much since the interpreter/debugger started working in Juno, so perhaps I'm not doing this the same way you are. So here's a screenshot:
![Revise_vscode](https://user-images.githubusercontent.com/1525481/64062672-24e1bd80-cbaf-11e9-8a2a-75e4ddcf9c50.png)

If you have more info about how you're running that script, I'm happy to look further.

If you haven't seen [this page](https://github.com/julia-vscode/julia-vscode/wiki/Known-issues-and-workarounds), might it help?

**Comment by clintonTE** [comment]
I use Juno mostly now too, but I love the code outline features of  VSCode, so I periodically try to get it going. 

I tried it the way you ran it and got the same result as you. I think the issue is the f5 - Start Debugging command in VSCode. When running the files straight using include statements, it seems to work. Here is how I have it set up after I hit f5 the first time:

![image](https://user-images.githubusercontent.com/25089431/64070564-f815c080-cc17-11e9-9e26-61370d3b66d5.png)

And then again after modifying the second file:
![image](https://user-images.githubusercontent.com/25089431/64070572-390dd500-cc18-11e9-8fc0-ad1271b71630.png)

I could but wrong, but I don't think I need to touch startup.jl since I am using Revise explicitly.

**Comment by timholy** [comment]
I can replicate this. The way Revise works is to [take over the REPL and call `revise` before executing each statement](https://github.com/timholy/Revise.jl/blob/26a55b75aeebb374908e7cfddc8f4436bf37e98e/src/Revise.jl#L939-L954). With IJulia and Atom we have hooks:
- https://github.com/timholy/Revise.jl/blob/26a55b75aeebb374908e7cfddc8f4436bf37e98e/src/Revise.jl#L1043-L1044
- https://github.com/timholy/Revise.jl/blob/26a55b75aeebb374908e7cfddc8f4436bf37e98e/src/Revise.jl#L1088-L1098

Is there something similar for VSCode we can do?

**Comment by pfitzseb** [comment]
Not without changes to the VSCode extension [here](https://github.com/julia-vscode/julia-vscode/blob/cddb43ba28248429f097f50f85b7c755a1c4c134/scripts/terminalserver/terminalserver.jl#L1289) and [here](https://github.com/julia-vscode/julia-vscode/blob/cddb43ba28248429f097f50f85b7c755a1c4c134/scripts/terminalserver/terminalserver.jl#L139) (and maybe elsewhere).

**Comment by timholy** [comment]
@pfitzseb should we close this?

**Comment by pfitzseb** [comment]
Yes, the extension now calls Revise, if it's loaded, before each code execution.

**Comment by jonathanlilly** [comment]
Revise is working for me in the REPL in VSCode, but not in a Jupyter Notebook running in VSCode unless I explicitly run revise().  Does anyone know if there is a way to get it to run the preexecute hook automatically?

**Comment by timholy** [comment]
Does the VSCode IJulia kernel use `preexecute_hooks`? If not, that's probably why.

Not a Revise issue, though. I've added #1066 just to ensure that everything needed by IJulia from Revise actually works; it does currently, and now we know it will keep working.

**Comment by jonathanlilly** [comment]
Ok, thank you!
